google/cloud/internal/async_streaming_write_rpc_timeout_test.cc

```diff
@@ -0,0 +1,331 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/async_streaming_write_rpc_timeout.h"
+#include "google/cloud/internal/make_status.h"
+#include "google/cloud/testing_util/async_sequencer.h"
+#include "google/cloud/testing_util/mock_async_streaming_write_rpc.h"
+#include "google/cloud/testing_util/mock_completion_queue_impl.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include <gmock/gmock.h>
+#include <memory>
+#include <string>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+namespace {
+
+using ::google::cloud::testing_util::AsyncSequencer;
+using ::google::cloud::testing_util::IsOkAndHolds;
+using ::google::cloud::testing_util::MockCompletionQueueImpl;
+using ::testing::_;
+using ::testing::ElementsAreArray;
+using ::testing::Return;
+
+using MockStream =
+    google::cloud::testing_util::MockAsyncStreamingWriteRpc<int, std::string>;
+using TestedStream = AsyncStreamingWriteRpcTimeout<int, std::string>;
+
+StatusOr<std::chrono::system_clock::time_point> CancelledTimer() {
+  return internal::CancelledError("test-only", GCP_ERROR_INFO());
+}
+
+TEST(AsyncStreamingWriteRpcTimeout, Cancel) {
+  AsyncSequencer<bool> sequencer;
+  auto mock = std::make_unique<MockStream>();
+  EXPECT_CALL(*mock, Cancel).Times(1);
+
+  auto constexpr kStartTimeout = std::chrono::milliseconds(2345);
+  auto constexpr kWriteTimeout = std::chrono::milliseconds(2345);
+
+  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
+  EXPECT_CALL(*mock_cq, MakeRelativeTimer).Times(0);
+
+  CompletionQueue cq(mock_cq);
+
+  TestedStream uut(cq, kStartTimeout, kWriteTimeout, std::move(mock));
+  uut.Cancel();
+}
+
+TEST(AsyncStreamingWriteRpcTimeout, StartSuccess) {
+  AsyncSequencer<bool> sequencer;
+  auto mock = std::make_unique<MockStream>();
+  EXPECT_CALL(*mock, Start).WillOnce([&] {
+    return sequencer.PushBack("Start");
+  });
+  EXPECT_CALL(*mock, Cancel).Times(0);
+
+  auto constexpr kStartTimeout = std::chrono::milliseconds(2345);
+  auto constexpr kWriteTimeout = std::chrono::milliseconds(2345);
+
+  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
+  EXPECT_CALL(*mock_cq,
+              MakeRelativeTimer(std::chrono::nanoseconds(kStartTimeout)))
+      .WillOnce([&](auto) {
+        return sequencer.PushBack("MakeRelativeTimer")
+            .then(
+                [](auto f) -> StatusOr<std::chrono::system_clock::time_point> {
+                  if (!f.get()) return CancelledTimer();
+                  return make_status_or(std::chrono::system_clock::now());
+                });
```

Comment:
nit: I find this boilerplate a little distracting. Consider:

```cc

StatusOr<std::chrono::system_clock::time_point> MakeTimerStatus(
    future<bool> expires) {
  if (expires.get()) return make_status_or(std::chrono::system_clock::now());
  return internal::CancelledError("test-only", GCP_ERROR_INFO());
}
```

```cc
return sequencer.PushBack("MakeRelativeTimer").then(MakeTimerStatus);
```

(I see this suggestion also applies to `google/cloud/internal/async_streaming_read_rpc_timeout_test.cc`. If you don't want to do it in this PR, that's fine.)

Reply:
Done.  Though we lose the correct like numbers from `GCP_ERROR_INFO()`. And now understanding each test requires going to the top of the file... 🤷